relayer/relays/beacon/header/header.go

```diff
@@ -259,32 +258,162 @@ func (h *Header) SyncHeaders(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
 
-	headerUpdate, err := h.SyncHeader(ctx, finalizedHeaderBlockRoot, prevSyncAggregate)
+	toHeaderUpdate, err := h.syncer.GetHeaderUpdate(toHeader)
 	if err != nil {
 		return err
 	}
 
-	prevSyncAggregate = headerUpdate.Block.Body.SyncAggregate
+	fromSlot := uint64(fromHeaderUpdate.Block.Slot)
+	toSlot := uint64(toHeaderUpdate.Block.Slot)
+	totalSlots := toSlot - fromSlot
 
-	for secondLastFinalizedHeader != blockRoot {
-		headerUpdate, err := h.SyncHeader(ctx, blockRoot, prevSyncAggregate)
+	log.WithFields(log.Fields{
+		"fromHeader": fromHeader,
+		"fromSlot":   fromSlot,
+		"fromEpoch":  h.syncer.ComputeEpochAtSlot(fromSlot),
+		"toHeader":   toHeader,
+		"toSlot":     toSlot,
+		"toEpoch":    h.syncer.ComputeEpochAtSlot(toSlot),
+		"totalSlots": totalSlots,
+	}).Info("starting to back-fill headers")
+
+	headersToSync := []syncer.HeaderUpdate{}
+
+	currentSlot := fromSlot + 1
+	for currentSlot <= toSlot {
+		epoch := h.syncer.ComputeEpochAtSlot(currentSlot)
+
+		// start of new epoch, sync headers of last epoch
+		if float64(epoch) >= (float64(currentSlot) / float64(h.syncer.SlotsInEpoch)) {
+			log.WithFields(log.Fields{
+				"epoch": h.syncer.ComputeEpochAtSlot(currentSlot) - 1,
+			}).Debug("syncing header in epoch")
+			for _, header := range headersToSync {
+				err := h.SyncHeader(ctx, header, toSlot-uint64(header.Block.Slot))
+				if err != nil {
+					return err
+				}
+			}
+
+			if len(headersToSync) > 0 {
+				err = h.sendLastBlockNumberMessage(ctx, uint64(headersToSync[len(headersToSync)-1].Block.Body.ExecutionPayload.BlockNumber), basicChannel, incentivizedChannel)
+				if err != nil {
+					return err
+				}
+			}
+
+			// new epoch, start with clean array
+			headersToSync = []syncer.HeaderUpdate{}
+		}
+
+		log.WithFields(log.Fields{
+			"slot": currentSlot,
+		}).Info("fetching header at slot")
+
+		headerUpdate, err := h.syncer.GetNextHeaderUpdateBySlot(currentSlot)
 		if err != nil {
 			return err
 		}
+		// To get the sync witness for the current synced header. This header
+		// will be used as the next update.
+		nextHeaderUpdate, err := h.syncer.GetNextHeaderUpdateBySlot(currentSlot + 1)
+		if err != nil {
+			return err
+		}
+
+		headerUpdate.SyncAggregate = nextHeaderUpdate.Block.Body.SyncAggregate
+
+		headersToSync = append(headersToSync, headerUpdate)
+		headerUpdate = nextHeaderUpdate
+
+		// last slot to be synced, sync headers
+		if currentSlot >= toSlot {
+			log.WithFields(log.Fields{
+				"epoch": h.syncer.ComputeEpochAtSlot(currentSlot),
+			}).Debug("syncing last set of headers in epoch")
+			for _, header := range headersToSync {
+				err := h.SyncHeader(ctx, header, toSlot-uint64(header.Block.Slot))
+				if err != nil {
+					return err
+				}
+			}
+
+			if len(headersToSync) > 0 {
+				err = h.sendLastBlockNumberMessage(ctx, uint64(headersToSync[len(headersToSync)-1].Block.Body.ExecutionPayload.BlockNumber), basicChannel, incentivizedChannel)
```

Comment:
Let's extract the last block number
```suggestion
				lastBlockNumber := uint64(headersToSync[len(headersToSync)-1].Block.Body.ExecutionPayload.BlockNumber)
				err = h.sendLastBlockNumberMessage(ctx, lastBlockNumber, basicChannel, incentivizedChannel)
```

Reply:
Addressed in https://github.com/Snowfork/snowbridge/pull/696/commits/dfc65eba1413eec725de2ec84cbd726519311b8b.